File: 프로그래머스/Level 2/타겟 넘버.js
/*
N: numbers.length
time: O(2^N)
space: O(N) for call stack
*/
function solution(numbers, target) {
	var answer = 0;
	dfs(numbers, 0, 0);

	function dfs(numbers, numIdx, sum) {
		if (numIdx === numbers.length) {
			if (sum === target) answer++;
			return;
		}
		dfs(numbers, numIdx + 1, sum + numbers[numIdx]);
		dfs(numbers, numIdx + 1, sum - numbers[numIdx]);
	}
	return answer;
}

File: 프로그래머스/python/Level 2/타겟 넘버.py
def solution(numbers, target):
    answer = 0
    
    def dfs(idx, sum):
        nonlocal answer
        if idx == len(numbers):
            if sum == target:
                answer += 1
            return
        dfs(idx+1, sum+numbers[idx])
        dfs(idx+1, sum-numbers[idx])
        
    dfs(0,0)
    return answer